src/utils/narrow.js

```diff
@@ -257,13 +257,82 @@ export function caseNarrowDefault<T>(
  * and `DraftsState`.
  */
 export function keyFromNarrow(narrow: Narrow): string {
-  return JSON.stringify(narrow);
+  // The ":s" bit in several of these is to keep them disjoint, out of an
+  // abundance of caution, from future keys that use numeric IDs.
+  return caseNarrow(narrow, {
+    stream: name => `stream:s:${name}`,
+    // '\x00' is the one character not allowed in Zulip stream names.
+    // (See `check_stream_name` in zulip.git:zerver/lib/streams.py.)
+    topic: (streamName, topic) => `topic:s:${streamName}\x00${topic}`,
+
+    pm: emails => `pm:s:${emails.join(',')}`,
+
+    home: () => 'all',
+    starred: () => 'starred',
+    mentioned: () => 'mentioned',
+    allPrivate: () => 'all-pm',
+    search: query => `search:${query}`,
+  });
 }
 
 /**
  * Parse a narrow previously encoded with keyFromNarrow.
  */
-export const parseNarrow = (narrowStr: string): Narrow => JSON.parse(narrowStr);
+export const parseNarrow = (narrowStr: string): Narrow => {
+  const makeError = () => new Error('parseNarrow: bad narrow');
+
+  const tag = /^.*?(?::|$)/.exec(narrowStr)?.[0] ?? '';
+  const rest = narrowStr.substr(tag.length);
+  // invariant: tag + rest === narrowStr
+  switch (tag) {
+    case 'stream:': {
+      if (!rest.startsWith('s:')) {
+        throw makeError();
+      }
+      const name = rest.substr('s:'.length);
+      return streamNarrow(name);
+    }
+
+    case 'topic:': {
+      // eslint-disable-next-line no-control-regex
+      const match = /^s:(.*?)\x00(.*)/s.exec(rest);
```

Comment:
It's not related to that character in the pattern, but it's for sort of an inverse kind of problem: it's so that the dots `.` match absolutely any character, even a newline if there somehow happens to be one in either the stream name or the topic. (The page you linked describes how `.` by default matches everything except four flavors of newline.)

As of this morning my draft of this change actually had a bug here because I didn't have the `/s`. I added it after writing the "awkward characters" part of the round-trip test, and finding that it failed on this change!

I'll add a line or so of a comment about this.

Reply:
(Meanwhile the `\x00` is there because it's the same character that `keyFromNarrow` uses as a delimiter. That choice is discussed in a short comment there.)